drv/lpc55-swd/src/main.rs

```diff
@@ -1058,21 +1297,626 @@ impl ServerImpl {
+        // running properly, there can only be one return at the end.
+
+        let digest = if prep(&mut need_undo).is_ok() {
+            self.do_measure_sp()
+        } else {
+            Err(())
+        };
+
+        // From here on, we're cleaning up and restarting the SP.
+
+        // It is very unlikely that an attached SP debug dongle would go
+        // active just as we are taking a measurement.
+        // If that happened, then JTAG DETECT will have its own notification
+        // and this task will perform an explicit attestation log reset.
+        // If there was any SWD problem for us, we may need to clean up
+        // ore or more of the steps that `prep` performed.
+
+        // If anything deviates from the happy-path, we will not record a valid measurement.
+
+        if need_undo & Undo::VC_CORERESET == Undo::VC_CORERESET {
+            // This should never happen. Don't believe any measurement we
+            // may have recorded.
+            ringbuf_entry!(Trace::IncompleteUndo(need_undo));
+            error = true;
+
+            if self
+                .dp_write_bitflags::<Demcr>(Demcr::from_bits_retain(0))
+                .is_ok()
+            {
+                need_undo &= !Undo::VC_CORERESET;
+                ringbuf_entry!(Trace::VcCoreReset(false));
+            } else {
+                ringbuf_entry!(Trace::DemcrWriteError);
+            }
+            if let Ok(r) = self.dp_read_bitflags::<Dhcsr>() {
+                ringbuf_entry!(Trace::Dhcsr(r));
+            }
+        }
+
+        // This read of DHCSR should never be useful given the code above.
+        // Read back to make sure that VC_CORERESET is clear.
+        // XXX remove
+        if let Ok(demcr) = self.dp_read_bitflags::<Demcr>() {
+            if (demcr & Demcr::VC_CORERESET) == Demcr::VC_CORERESET {
+                error = true;
+                need_undo |= Undo::VC_CORERESET;
+                ringbuf_entry!(Trace::BadDemcr(demcr));
+            }
+        } else {
+            ringbuf_entry!(Trace::DemcrReadError);
+        }
+
+        // Unless `prep` failed, this will always be needed.
+        if need_undo & Undo::DEBUGEN == Undo::DEBUGEN {
+            if self.dp_write_bitflags::<Dhcsr>(Dhcsr::end_debug()).is_ok() {
+                need_undo &= !Undo::DEBUGEN;
+            } else {
+                ringbuf_entry!(Trace::DhcsrWriteError);
+            }
+        } else {
+            ringbuf_entry!(Trace::Never);
+        }
+
+        // This should always be needed
+        if need_undo & Undo::SWD == Undo::SWD {
+            self.swd_finish();
+            need_undo &= !Undo::SWD;
+        } else {
+            ringbuf_entry!(Trace::Never);
+            error = true;
+        }
+
+        if !need_undo.is_empty() {
+            ringbuf_entry!(Trace::IncompleteUndo(need_undo));
+            error = true;
+        }
+
+        // The SP is still halted.
+        // Get it running again by toggling its RESET pin.
```

Comment:
It would also be nice to leave a comment here explaining that the caller is responsible for clearing the pending IRQ, so we don't measure it again.

Reply:
I refactored the pending interrupt management.